File: src/shared/utils/modpack.test.ts
import { describe, it, expect } from 'vitest'
import { resolveModpackLoader, isUnsafeModpackPath, parseModrinthModpackSlug } from './modpack'

describe('resolveModpackLoader', () => {
  it('resolves Fabric packs', () => {
    expect(resolveModpackLoader({ minecraft: '1.20.1', 'fabric-loader': '0.15.7' })).toEqual({
      minecraftVersion: '1.20.1',
      loader: 'fabric',
      loaderVersion: '0.15.7'
    })
  })

  it('resolves NeoForge and Forge and Quilt', () => {
    expect(resolveModpackLoader({ minecraft: '1.21', neoforge: '21.0.5' }).loader).toBe('neoforge')
    expect(resolveModpackLoader({ minecraft: '1.20.1', forge: '47.2.0' }).loader).toBe('forge')
    expect(resolveModpackLoader({ minecraft: '1.20.1', 'quilt-loader': '0.23' }).loader).toBe('quilt')
  })

  it('returns a null loader for a vanilla pack', () => {
    expect(resolveModpackLoader({ minecraft: '1.20.1' })).toEqual({
      minecraftVersion: '1.20.1',
      loader: null,
      loaderVersion: null
    })
  })
})

describe('isUnsafeModpackPath', () => {
  it('accepts normal relative paths', () => {
    expect(isUnsafeModpackPath('mods/sodium.jar')).toBe(false)
    expect(isUnsafeModpackPath('config/foo/bar.toml')).toBe(false)
  })

  it('rejects traversal, absolute, and drive paths', () => {
    expect(isUnsafeModpackPath('../evil.jar')).toBe(true)
    expect(isUnsafeModpackPath('mods/../../evil.jar')).toBe(true)
    expect(isUnsafeModpackPath('/etc/passwd')).toBe(true)
    expect(isUnsafeModpackPath('C:\\Windows\\system32')).toBe(true)
    expect(isUnsafeModpackPath('')).toBe(true)
  })
})

describe('parseModrinthModpackSlug', () => {
  it('extracts the slug from a modpack URL', () => {
    expect(parseModrinthModpackSlug('https://modrinth.com/modpack/fabulously-optimized')).toBe(
      'fabulously-optimized'
    )
    expect(
      parseModrinthModpackSlug('https://modrinth.com/modpack/create-fabric/version/abc123')
    ).toBe('create-fabric')
  })

  it('accepts a bare slug', () => {
    expect(parseModrinthModpackSlug('  cobblemon  ')).toBe('cobblemon')
  })

  it('rejects empty input and non-modrinth URLs', () => {
    expect(parseModrinthModpackSlug('')).toBeNull()
    expect(parseModrinthModpackSlug('https://curseforge.com/minecraft/modpacks/rlcraft')).toBeNull()
  })
})

File: daemon/internal/modpack/modpack_test.go
package modpack

import "testing"

func TestResolveLoader(t *testing.T) {
	mc, lt, lv := resolveLoader(map[string]string{"minecraft": "1.20.1", "fabric-loader": "0.15.7"})
	if mc != "1.20.1" || lt != "fabric" || lv != "0.15.7" {
		t.Fatalf("fabric resolve = %q %q %q", mc, lt, lv)
	}
	if _, lt, _ := resolveLoader(map[string]string{"minecraft": "1.21", "neoforge": "21.0.5"}); lt != "neoforge" {
		t.Fatalf("neoforge not resolved")
	}
	if _, lt, _ := resolveLoader(map[string]string{"minecraft": "1.20.1"}); lt != "" {
		t.Fatalf("vanilla should have no loader")
	}
}

func TestIsUnsafeRel(t *testing.T) {
	for _, p := range []string{"../evil", "/etc/passwd", "C:\\x", "mods/../../x", ""} {
		if !isUnsafeRel(p) {
			t.Errorf("expected %q unsafe", p)
		}
	}
	for _, p := range []string{"mods/sodium.jar", "config/a/b.toml"} {
		if isUnsafeRel(p) {
			t.Errorf("expected %q safe", p)
		}
	}
}
